controllers/vmware/vspheremachinetemplate_controller_test.go

```diff
@@ -0,0 +1,185 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vmware
+
+import (
+	"context"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	vmoprv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	vmwarev1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
+)
+
+func Test_vSphereMachineTemplateReconciler_Reconcile(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+
+	scheme := runtime.NewScheme()
+	g.Expect(corev1.AddToScheme(scheme)).To(Succeed())
+	g.Expect(vmwarev1.AddToScheme(scheme)).To(Succeed())
+	g.Expect(vmoprv1.AddToScheme(scheme)).To(Succeed())
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-namespace",
+		},
+	}
+
+	tests := []struct {
+		name                   string
+		vSphereMachineTemplate *vmwarev1.VSphereMachineTemplate
+		objects                []client.Object
+		wantErr                bool
+		wantStatus             *vmwarev1.VSphereMachineTemplateStatus
+	}{
+		{
+			name:                   "object does not exist",
+			vSphereMachineTemplate: nil,
+			objects:                []client.Object{},
+			wantErr:                false,
+			wantStatus:             nil,
+		},
+		{
+			name:                   "VirtualMachineClass does not exist",
+			vSphereMachineTemplate: vSphereMachineTemplate(namespace.Name, "no-class", "not-existing-class", nil),
+			objects:                []client.Object{},
+			wantErr:                true,
+			wantStatus:             nil,
+		},
+		{
+			name:                   "VirtualMachineClass does exist but has no data",
+			vSphereMachineTemplate: vSphereMachineTemplate(namespace.Name, "with-class", "vm-class", nil),
+			objects: []client.Object{
+				virtualMachineClass(namespace.Name, "vm-class", nil),
+			},
+			wantErr:    false,
+			wantStatus: &vmwarev1.VSphereMachineTemplateStatus{},
+		},
+		{
+			name:                   "VirtualMachineClass does exist and has cpu and memory set",
+			vSphereMachineTemplate: vSphereMachineTemplate(namespace.Name, "with-class", "vm-class", nil),
+			objects: []client.Object{
+				virtualMachineClass(namespace.Name, "vm-class", &vmoprv1.VirtualMachineClassHardware{Cpus: 1, Memory: *resource.NewQuantity(1024, resource.DecimalSI)}),
```

Comment:
Stupid question, but we know that in our use case CPU & memory are set in these fields right? (and not in some other way)

Reply:
Yes, example from a recent testbed:

```
apiVersion: vmoperator.vmware.com/v1alpha2
kind: VirtualMachineClass
metadata:
  creationTimestamp: "2024-08-21T14:22:41Z"
  generation: 1
  name: best-effort-2xlarge
  namespace: svc-tkg-domain-c52
  resourceVersion: "1713184"
  uid: 325a1703-5848-4539-b0a9-c04edf828df6
spec:
  configSpec:
    _typeName: VirtualMachineConfigSpec
    memoryMB: 65536
    numCPUs: 8
  hardware:
    cpus: 8
    devices: {}
    instanceStorage: {}
    memory: 64Gi
  policies:
    resources:
      limits:
        cpu: "0"
        memory: "0"
      requests:
        cpu: "0"
        memory: "0"
```

No idea why they duplicate the hw and configSpec part but 🤷

Reply:
Thx!